src/protocol/sort/multi_bit_permutation.rs

```diff
@@ -0,0 +1,256 @@
+            })
+            .collect::<Vec<_>>();
+        initial_state = out
+            .last()
+            .unwrap_or_else(|| panic!("We should never reach here"))
+            .1
+            .clone();
+        sop_inputs_transposed.push(out);
+    }
+
+    // Take sum of products of output of equality check and accumulated sum
+    let mut permutation_futures = Vec::with_capacity(num_records);
+    for rec in 0..num_records {
+        let ctx_sop = ctx.narrow(&Sop);
+        let mut sop_inputs = Vec::with_capacity(num_possible_bit_values);
+        for sop_input_transposed in sop_inputs_transposed.iter().take(num_possible_bit_values) {
+            sop_inputs.push((&sop_input_transposed[rec].0, &sop_input_transposed[rec].1));
+        }
+        permutation_futures.push(async move {
+            ctx_sop
+                .sum_of_products(RecordId::from(rec), sop_inputs.as_slice())
+                .await
+        });
+    }
+    try_join_all(permutation_futures).await
+}
+
+/// For a given `idx` check if each of the record has same value as idx.
+/// Steps
+/// 1. Get bit representation of `idx`
+/// 2. Calculate equality check
+///   i. keep record bit if `idx` bit is 1
+///   ii.toggle record bit if idx bit is 0 (Done by taking `share_of_one` - value)
+/// 3. Multiply equality checks for all bits per record - in clear per record this will be 1 only for 1 index while 0 for all others
+async fn get_bit_equality_checkers<F, C, S>(
+    idx: usize,
+    input: &[Vec<S>],
+    ctx: C,
+) -> Result<Vec<S>, Error>
+where
+    F: Field,
+    C: Context<F, Share = S>,
+    S: SecretSharing<F>,
+{
+    let num_multi_bits = input.len();
+    let num_records = input[0].len();
+
+    let mut equality_check_futures = Vec::with_capacity(num_records);
+
+    let j_bit_representation = get_binary_from_int(idx, num_multi_bits);
+
+    for rec in 0..num_records {
+        let ctx_across_bits = ctx.narrow(&MultiplyAcrossBits);
+        let share_of_one = ctx_across_bits.share_of_one();
+
+        let mut bits_for_record = Vec::with_capacity(num_multi_bits);
+        for jth_bits in input.iter().take(num_multi_bits) {
+            bits_for_record.push(jth_bits[rec].clone());
+        }
+        let mut mult_input = Vec::with_capacity(num_multi_bits);
+        for jth_bit in j_bit_representation.iter().take(num_multi_bits) {
+            if *jth_bit {
+                mult_input.push(bits_for_record.remove(0));
+            } else {
+                mult_input.push(-bits_for_record.remove(0) + &share_of_one);
+            }
+        }
+        equality_check_futures.push(async move {
+            // multiply all mult_input for this j for each record => f(j)
+            multiply_all_shares(
+                ctx_across_bits,
+                RecordId::from(idx * num_records + rec),
+                mult_input.as_slice(),
+            )
+            .await
+        });
+    }
+    try_join_all(equality_check_futures).await
```

Comment:
`try_join_all` takes an iterator, you probably don't need to keep futures inside a vector. `try_join_all(num_records.map(..)).await?` will likely be more efficient